chiron/multistate.py

```diff
@@ -0,0 +1,968 @@
+import copy
+import time
+from typing import List, Optional
+from chiron.states import SamplerState, ThermodynamicState
+import datetime
+from loguru import logger as log
+import numpy as np
+from openmmtools.utils import time_it, with_timer
+from chiron.neighbors import NeighborListNsqrd
+from openmm import unit
+from chiron.mcmc import MCMCMove
+from openmmtools.multistate import MultiStateReporter
+
+
+class MultiStateSampler(object):
+    """
+    Base class for samplers that sample multiple thermodynamic states using
+    one or more replicas.
+
+    This base class provides a general simulation facility for multistate from multiple
+    thermodynamic states, allowing any set of thermodynamic states to be specified.
+    If instantiated on its own, the thermodynamic state indices associated with each
+    state are specified and replica mixing does not change any thermodynamic states,
+    meaning that each replica remains in its original thermodynamic state.
+
+    Parameters
+    ----------
+    mcmc_moves : MCMCMove or list of MCMCMove, optional
+        The MCMCMove used to propagate the thermodynamic states. If a list of MCMCMoves,
+        they will be assigned to the correspondent thermodynamic state on
+        creation. If None is provided, Langevin dynamics with 2fm timestep, 5.0/ps collision rate,
+        and 500 steps per iteration will be used.
+    number_of_iterations : int or infinity, optional, default: 1
+        The number of iterations to perform.
+
+    locality : int > 0, optional, default None
+        If None, the energies at all states will be computed for every replica each iteration.
+        If int > 0, energies will only be computed for states ``range(max(0, state-locality), min(n_states, state+locality))``.
+
+    Attributes
+    ----------
+    n_replicas
+    n_states
+    iteration
+    mcmc_moves
+    sampler_states
+    metadata
+    is_completed
+    """
+
+    def __init__(self, mcmc_moves=None, number_of_iterations=1, locality=None):
+        # These will be set on initialization. See function
+        # create() for explanation of single variables.
+        self._thermodynamic_states = None
+        self._unsampled_states = None
+        self._sampler_states = None
+        self._replica_thermodynamic_states = None
+        self._iteration = None
+        self._energy_thermodynamic_states = None
+        self._neighborhoods = None
+        self._energy_unsampled_states = None
+        self._n_accepted_matrix = None
+        self._n_proposed_matrix = None
+        self._reporter = None
+        self._metadata = None
+        self._timing_data = dict()
+
+        # Handling default propagator.
```

Comment:
We can simplify this if we require `mcmc_moves` be specified as an argument. That may not be a bad idea here.